mlir/lib/QEC/Transforms/CountPPMSpecs.cpp

```diff
@@ -0,0 +1,124 @@
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+using namespace catalyst::qec;
+using json = nlohmann::json;
+
+namespace catalyst {
+namespace qec {
+
+#define GEN_PASS_DEF_COUNTPPMSPECSPASS
+#define GEN_PASS_DECL_COUNTPPMSPECSPASS
+#include "QEC/Transforms/Passes.h.inc"
+
+struct CountPPMSpecsPass : public impl::CountPPMSpecsPassBase<CountPPMSpecsPass> {
+    using CountPPMSpecsPassBase::CountPPMSpecsPassBase;
+
+    llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>>
+    countLogicalQubit(Operation *op,
+                      llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>> PPMSpecs)
+    {
+        uint64_t numQubits = cast<quantum::AllocOp>(op).getNqubitsAttr().value_or(0);
+        assert(numQubits != 0 && "PPM specs with dynamic number of qubits is not implemented");
+        auto parentFuncOp = op->getParentOfType<func::FuncOp>();
+        PPMSpecs[parentFuncOp.getName()]["num_logical_qubits"] = numQubits;
+        return PPMSpecs;
+    }
+
+    llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>>
+    countPPM(Operation *op, llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>> PPMSpecs)
+    {
+        auto parentFuncOp = op->getParentOfType<func::FuncOp>();
+        PPMSpecs[parentFuncOp.getName()]["num_of_ppm"]++;
+        return PPMSpecs;
+    }
+
+    llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>>
+    countPPR(Operation *op, llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>> PPMSpecs,
+             llvm::BumpPtrAllocator *stringAllocator)
+    {
+        int16_t rotationKind =
+            cast<qec::PPRotationOp>(op).getRotationKindAttr().getValue().getZExtValue();
+        auto PauliProductAttr = cast<qec::PPRotationOp>(op).getPauliProductAttr();
+        auto parentFuncOp = op->getParentOfType<func::FuncOp>();
+        StringRef funcName = parentFuncOp.getName();
+        if (rotationKind) {
+            llvm::StringSaver saver(*stringAllocator);
+            StringRef numRotationKindKey =
+                saver.save("num_pi" + std::to_string(abs(rotationKind)) + "_gates");
+            StringRef maxWeightRotationKindKey =
+                saver.save("max_weight_pi" + std::to_string(abs(rotationKind)));
+            PPMSpecs[funcName][numRotationKindKey]++;
+            PPMSpecs[funcName][maxWeightRotationKindKey] =
+                std::max(PPMSpecs[funcName][maxWeightRotationKindKey],
+                         static_cast<int>(PauliProductAttr.size()));
+        }
+        return PPMSpecs;
+    }
+    void printSpecs()
+    {
+        llvm::BumpPtrAllocator stringAllocator;
+        llvm::DenseMap<StringRef, llvm::DenseMap<StringRef, int>> PPMSpecs;
+        // StringRef funcName;
+        // Walk over all operations in the IR (could be ModuleOp or FuncOp)
+        getOperation()->walk([&](Operation *op) {
+            // Skip top-level container ops if desired
+            if (isa<ModuleOp>(op))
+                return;
+
```

Comment:
```suggestion
            if (isa<ModuleOp>(op)) {
                return;
            }
```
https://github.com/PennyLaneAI/guidance-docs/blob/master/development/cpp-dev-guidelines.md#brace-style

> Always favour brace use, even for short function definitions/control-flow/etc.
> Example:
> 
> // Allowed
> for(std::size_t loop_var = 0; loop_var < 10; loop_var++){
>     my_function(loop_var);
> }
> Use the above, rather than
> 
> // Disallowed
> for(std::size_t loop_var = 0; loop_var < 10; loop_var++)
>     my_function(loop_var);

Reply:
https://github.com/PennyLaneAI/catalyst/commit/f063e5217546aa827c62ec3bd0c68fa970de9d25